aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -387,6 +383,90 @@ impl AptosVM {
         }
     }
 
+    fn finish_aborted_transaction(
+        &self,
+        gas_meter: &mut impl AptosGasMeter,
+        txn_data: &TransactionMetadata,
+        resolver: &impl AptosMoveResolver,
+        log_context: &AdapterLogSchema,
+        change_set_configs: &ChangeSetConfigs,
+    ) -> Result<(VMChangeSet, FeeStatement), VMStatus> {
+        if txn_data.fee_payer().is_some()
+            && txn_data.sequence_number == 0
+            && self
+                .vm_impl
+                .get_features()
+                .is_enabled(FeatureFlag::SPONSORED_AUTOMATIC_ACCOUNT_CREATION)
+        {
+            // Clear side effects: create new session and clear refunds from fee statement.
+            let mut session = self
+                .vm_impl
+                .new_session(resolver, SessionId::run_on_abort(txn_data));
+
+            create_account_if_does_not_exist(&mut session, gas_meter, txn_data.sender())
+                .map_err(expect_no_verification_errors)
+                .or_else(|err| {
+                    expect_only_successful_execution(
+                        err,
+                        &format!("{:?}::{}", ACCOUNT_MODULE, CREATE_ACCOUNT_IF_DOES_NOT_EXIST),
+                        log_context,
+                    )
+                })?;
+
+            let mut change_set = session.finish(change_set_configs)?;
+            if let Err(err) = self.charge_change_set(&mut change_set, gas_meter, txn_data) {
```

Comment:
Shall we add a flag to not record deposits on the new slots created (so no refund for these slots)? Makes me feel safer.

Reply:
why do that? if we can give them, we should

Reply:
So we will never create a bug where you can get more refund then you've paid. I know you defend against that by requiring more that two slots worth of it upfront, but I'm not feeling safe -- one can accidentally remove the requirement.

Reply:
can we somehow make this more robust, either:
- make sure that charge we are missing here is below what we already charged, or
- refund what we already charged, charge here expecting no failures, and then try to charge back how much we refunded?

Reply:
I am wondering if we can at least capture these cases in tests, e.g. have a test to check the refund is < what was paid? In case some checks are removed we will see something is off

Reply:
Assuming you landed on the idea that the validation function should test (2* storage + 10*gas_price) which I think is much, much more expensive then it is to create an account. I think that is acceptable... we should be willing to charge more for aborted transactions.